include/ppx/knob.h

```diff
@@ -0,0 +1,284 @@
+    //   - mCallback is triggered
+    void SetValue(bool newVal, bool updateDefault = false);
+
+private:
+    bool mValue;
+    bool mDefaultValue;
+
+    // mCallback will be called on the new bool value every time it is toggled in the UI, or updated with SetValue()
+    std::function<void(bool)> mCallback;
+};
+
+// KnobIntSlider is an int knob that will be displayed as a slider in the UI
+// ImGui sliders can also become input boxes with ctrl + right click
+class KnobIntSlider
+    : public Knob
+{
+public:
+    KnobIntSlider(const std::string& displayName, const std::string& flagName, const std::string& flagDesc, int defaultValue, int minValue, int maxValue, std::function<void(int)> callback = NULL)
+        : Knob(displayName, flagName, flagDesc)
+    {
+        if (minValue >= maxValue) {
+            PPX_LOG_FATAL(flagName << " KnobIntSlider: slider invalid range " << minValue << "~" << maxValue);
+        }
+        mMinValue = minValue;
+        mMaxValue = maxValue;
+        mCallback = callback;
+        SetValue(defaultValue, true);
+    }
+
+    void Draw() override;
+    void ResetToDefault() override { SetValue(mDefaultValue); }
+
+    // FlagText format:
+    // --flag_name <mMinValue~mMaxValue> : flag description
+    std::string FlagText() const override;
+
+    // Updates mValue and mDefaultValue, expected flag format:
+    // --flag_name <int>
+    Result UpdateFromFlag(const CliOptions& opts) override;
+
+    int GetValue() const { return mValue; }
+
+    // Used for when mValue needs to be updated outside of UI
+    // Returns success if newVal is valid, fails otherwise
+    // If newVal is valid:
+    // - If updateDefault, mDefaultValue is updated
+    // - If newVal is different from mValue:
+    //   - mValue is updated
+    //   - mCallback is triggered
+    Result SetValue(int newVal, bool updateDefault = false);
+
+private:
+    int mValue;
+    int mDefaultValue;
+
+    // mValue will be clamped to the mMinValue to mMaxValue range, inclusive
+    int mMinValue;
+    int mMaxValue;
+
+    // mCallback will be called on the new int value every time the slider is deselected in the UI, or updated with SetIntValue()
+    std::function<void(int)> mCallback;
+};
+
+// KnobStrDropdown is an int knob that will be displayed as a dropdown in the UI
+// The knob stores the index of a selected choice from a list of allowed strings
+class KnobStrDropdown
+    : public Knob
+{
+public:
+    KnobStrDropdown(const std::string& displayName, const std::string& flagName, const std::string& flagDesc, int defaultIndex, const std::vector<std::string>& choices, std::function<void(int)> callback = NULL)
+        : Knob(displayName, flagName, flagDesc)
+    {
+        if (choices.size() < 1) {
+            PPX_LOG_FATAL(flagName << " KnobStrDropdown: too few allowed choices");
+        }
+        mChoices  = choices;
+        mCallback = callback;
+        auto res  = SetIndex(defaultIndex, true);
+        if (res != SUCCESS)
```

Comment:
Really believe this logic should not be in the knob framework.
I believe we should assert here, when the compile-time constants used are bad.
But when we set the default through a flag, the check should probably be done outside, when validating command-line parameters.